### PR TITLE
AEA-1208 Fix upgrade command to eject packages when needed (#2133)

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Fix upgrade command to eject packages when needed (#2133)

## Fixes

n/a

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/cli/upgrade.py [hunk 1]
```diff
@@ -103,23 +110,42 @@ def upgrade_project(ctx: Context) -> None:  # pylint: disable=unused-argument
...
+        old_aea_version_specifier = ctx.agent_config.aea_version
         new_aea_version_specifier = compute_specifier_from_version(version)
+        click.echo(
+            f"Upgrading version specifier from {old_aea_version_specifier} to {new_aea_version_specifier}."
+        )
         ctx.agent_config.aea_version = new_aea_version_specifier
         ctx.dump_agent_config()
 
+    eject_helper = InteractiveEjectHelper(ctx, agent_items, interactive=interactive)
+    eject_helper.get_latest_versions()
```
Comment (marcofavorito) [inline aea/cli/upgrade.py:135]: we have to get the latest versions of all the packages, this makes the command a bit slow

Comment (marcofavorito) [reply aea/cli/upgrade.py:135]: we can refine by filtering non-vendor packages, but not if package is in requirements of another package, as done before (or not?)

**Review by solarw — APPROVED** [review]
LGTM

File: aea/cli/upgrade.py [hunk 2]
```diff
@@ -103,23 +110,42 @@ def upgrade_project(ctx: Context) -> None:  # pylint: disable=unused-argument
...
+        click.echo("Everything is already up to date!")
+        return
+    if not eject_helper.can_eject():
+        click.echo("Abort.")
+        return
+    eject_helper.eject()
+
     for package_id, deps in agent_items.items():
+        if package_id in eject_helper.to_eject:
+            continue
```
Comment (DavidMinarsch) [inline aea/cli/upgrade.py:146]: can we isolate some code blocks here and give them proper names; for instance starting at 144, what is the code doing? would be much easier to read if this was better structured

File: aea/cli/upgrade.py [hunk 3]
```diff
@@ -316,6 +357,115 @@ def add_item(self) -> None:
...
+        Ask the user permission for ejection of a package.
+
+        :param package_id: the package id.
+        :param dependencies_to_upgrade: the dependencies to upgrade.
+        :return: True or False, depending on the answer of the user.
+        """
+        package_type = str(package_id.package_type).capitalize()
+        message = (
+            f"{package_type} {package_id.public_id} prevents the upgrade of "
+            f"the following non-vendor packages:\n"
```
Comment (DavidMinarsch) [inline aea/cli/upgrade.py:460]:
```suggestion
            f"the following vendor packages:\n"
```

File: aea/cli/upgrade.py [hunk 4]
```diff
@@ -316,6 +357,115 @@ def add_item(self) -> None:
...
+            f"{pprint.pformat(dependencies_to_upgrade)}\n"
+            f"as there isn't an update available. "
```
Comment (DavidMinarsch) [inline aea/cli/upgrade.py:462]:
```suggestion
            f"as there isn't a compatible version available on the AEA registry. "
```

**Review by DavidMinarsch — COMMENTED** [review]
Looks mostly good to me. A few observations in comments and some more here:
- Run `aea fetch Aeolist/simple_weather_buyer:0.2.0` then `cd simple_weather_buyer` then `aea upgrade`. Command works as expected, however some issues remain: a) author is not updated in `aea-config`. Every time we upgrade we are changing the agent so the author needs to be changed if it doesn't match. Version needs to be reset if doesn't match (same as for eject). b) Eject command does not update the version ranges in the ejected packages.

File: aea/cli/upgrade.py (hunk 1 above)
Comment (DavidMinarsch) [reply aea/cli/upgrade.py:135]: I am ok with how it is. We don't care if the command is not that fast.

File: aea/configurations/constants.py [hunk 5]
```diff
@@ -59,7 +59,6 @@
 DEFAULT_CONNECTION_CONFIG_FILE = "connection.yaml"
 DEFAULT_CONTRACT_CONFIG_FILE = "contract.yaml"
 DEFAULT_PROTOCOL_CONFIG_FILE = "protocol.yaml"
-DEFAULT_LICENSE = "Apache-2.0"
```
Comment (marcofavorito) [inline aea/configurations/constants.py:62]: defined above without usage

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**